shared/ssz/hash.go

```diff
@@ -0,0 +1,211 @@
+	case kind == reflect.Slice && typ.Elem().Kind() == reflect.Uint8 ||
+		kind == reflect.Array && typ.Elem().Kind() == reflect.Uint8:
+		return hashedEncoding, nil
+	case kind == reflect.Slice || kind == reflect.Array:
+		return makeSliceHasher(typ)
+	case kind == reflect.Struct:
+		return makeStructHasher(typ)
+	case kind == reflect.Ptr:
+		return makePtrHasher(typ)
+	default:
+		return nil, fmt.Errorf("type %v is not hashable", typ)
+	}
+}
+
+func getEncoding(val reflect.Value) ([]byte, error) {
+	utils, err := cachedSSZUtilsNoAcquireLock(val.Type())
+	if err != nil {
+		return nil, err
+	}
+	buf := &encbuf{}
+	if err = utils.encoder(val, buf); err != nil {
+		return nil, err
+	}
+	writer := new(bytes.Buffer)
+	if err = buf.toWriter(writer); err != nil {
+		return nil, err
+	}
+	return writer.Bytes(), nil
+}
+
+func hashedEncoding(val reflect.Value) ([]byte, error) {
+	encoding, err := getEncoding(val)
+	if err != nil {
+		return nil, err
+	}
+	output := hashutil.Hash(encoding)
+	return output[:], nil
+}
+
+func makeSliceHasher(typ reflect.Type) (hasher, error) {
+	elemSSZUtils, err := cachedSSZUtilsNoAcquireLock(typ.Elem())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ssz utils: %v", err)
+	}
+	hasher := func(val reflect.Value) ([]byte, error) {
+		var elemHashList [][]byte
+		for i := 0; i < val.Len(); i++ {
+			elemHash, err := elemSSZUtils.hasher(val.Index(i))
+			if err != nil {
+				return nil, fmt.Errorf("failed to hash element of slice/array: %v", err)
+			}
+			elemHashList = append(elemHashList, elemHash)
+		}
+		output, err := merkleHash(elemHashList)
+		if err != nil {
+			return nil, fmt.Errorf("failed to calculate merkle hash of element hash list: %v", err)
+		}
+		return output, nil
+	}
+	return hasher, nil
+}
+
+func makeStructHasher(typ reflect.Type) (hasher, error) {
+	fields, err := structFields(typ)
+	if err != nil {
+		return nil, err
+	}
+	hasher := func(val reflect.Value) ([]byte, error) {
+		concatElemHash := make([]byte, 0)
+		for _, f := range fields {
+			elemHash, err := f.sszUtils.hasher(val.Field(f.index))
+			if err != nil {
+				return nil, fmt.Errorf("failed to hash field of struct: %v", err)
+			}
+			concatElemHash = append(concatElemHash, elemHash...)
+		}
+		result := hashutil.Hash(concatElemHash)
+		return result[:], nil
+	}
+	return hasher, nil
+}
+
+// Notice: Currently we don't support nil pointer:
+// - Input for encoding must not contain nil pointer
+// - Output for decoding will never contain nil pointer
+// (Not to be confused with empty slice. Empty slice is supported)
+func makePtrHasher(typ reflect.Type) (hasher, error) {
+	elemSSZUtils, err := cachedSSZUtilsNoAcquireLock(typ.Elem())
+	if err != nil {
+		return nil, err
+	}
+
+	hasher := func(val reflect.Value) ([]byte, error) {
+		if val.IsNil() {
+			return nil, errors.New("nil is not supported")
+		}
+		return elemSSZUtils.hasher(val.Elem())
+	}
+	return hasher, nil
+}
+
+func merkleHash(list [][]byte) ([]byte, error) {
```

Comment:
Hey @houjieth can we have some more detailed comments of the internal logic at the top of this function for easier readability?

Reply:
Sure. Just added some comments.

Reply:
can you also add a link of the official spec?

Reply:
Yes. Just added the link.